flow/table_test.go

```diff
@@ -0,0 +1,300 @@
+	checkFlowTable(t, ft)
+	if "1 flows" != ft.String() {
+		t.Error("We should got only 1 flow")
+	}
+	flow = &Flow{}
+	flow.UUID = "4567"
+	flows = append(flows, flow)
+	ft.Update(flows)
+	checkFlowTable(t, ft)
+	ft.Update(flows)
+	checkFlowTable(t, ft)
+	if "2 flows" != ft.String() {
+		t.Error("We should got only 2 flows")
+	}
+	return ft
+}
+func NewFlowTableComplex(t *testing.T) *FlowTable {
+	ft := NewFlowTable()
+	flows := generateFlows(t, ft)
+	ft = ft.NewFlowTableFromFlows(flows)
+	checkFlowTable(t, ft)
+	return ft
+}
+
+func TestNewFlowTable(t *testing.T) {
+	ft := NewFlowTable()
+	if ft == nil {
+		t.Error("new FlowTable return null")
+	}
+}
+func TestFlowTable_String(t *testing.T) {
+	ft := NewFlowTable()
+	if "0 flows" != ft.String() {
+		t.Error("FlowTable too big")
+	}
+	ft = NewFlowTableSimple(t)
+	if "2 flows" != ft.String() {
+		t.Error("FlowTable too big")
+	}
+}
+func TestFlowTable_Update(t *testing.T) {
+	ft := NewFlowTableSimple(t)
+	/* simulate a collision */
+	f := &Flow{}
+	ft.table["789"] = f
+	ft.table["789"].UUID = "78910"
+	f = &Flow{}
+	f.UUID = "789"
+	ft.Update([]*Flow{f})
+	NewFlowTableComplex(t)
+}
+
+type MyTestFlowCounter struct {
+	NbFlow int
+}
+
+func (fo *MyTestFlowCounter) expireCallback(f *Flow) {
+	fo.NbFlow++
+}
+
+func TestFlowTable_expire(t *testing.T) {
+	const MaxInt64 = int64(^uint64(0) >> 1)
+	ft := NewFlowTableComplex(t)
+
+	fc := MyTestFlowCounter{}
+	beforeNbFlow := fc.NbFlow
+	ft.expire(fc.expireCallback, 0)
+	afterNbFlow := fc.NbFlow
+	if beforeNbFlow != 0 && afterNbFlow != 0 {
+		t.Error("we should not expire a flow")
+	}
+
+	fc = MyTestFlowCounter{}
+	beforeNbFlow = fc.NbFlow
+	ft.expire(fc.expireCallback, MaxInt64)
+	afterNbFlow = fc.NbFlow
+	if beforeNbFlow != 0 && afterNbFlow != 10 {
+		t.Error("we should expire all flows")
+	}
+}
+
+func TestFlowTable_AsyncExpire(t *testing.T) {
+	t.Skip()
+}
+
+func TestFlowTable_IsExist(t *testing.T) {
+	ft := NewFlowTableSimple(t)
+	flow := &Flow{}
+	flow.UUID = "1234"
+	if ft.IsExist(flow) == false {
+		t.Fail()
+	}
+	flow.UUID = "12345"
+	if ft.IsExist(flow) == true {
+		t.Fail()
+	}
+}
+
+func TestFlowTable_GetFlow(t *testing.T) {
+	ft := NewFlowTableComplex(t)
+	flows := generateFlows(t, ft)
+	if len(flows) != 10 {
+		t.Error("missing some flows ", len(flows))
+	}
+	forgeEthIPTCP(t, int64(1234))
+	_, new := ft.GetFlow("abcd")
+	if !new {
+		t.Error("Collision in the FlowTable, should be new")
+	}
+	forgeEthIPTCP(t, int64(1234))
+	_, new = ft.GetFlow("abcd")
+	if new {
+		t.Error("Collision in the FlowTable, should be an update")
+	}
+	forgeEthIPTCP(t, int64(1234))
+	_, new = ft.GetFlow("abcde")
+	if !new {
+		t.Error("Collision in the FlowTable, should be a new flow")
+	}
+}
+
+func isJSON(s string) bool {
+	var js map[string]interface{}
+	return json.Unmarshal([]byte(s), &js) == nil
+}
+
+func TestFlowTable_JSONFlowConversationEthernetPath(t *testing.T) {
+	ft := NewFlowTableComplex(t)
+	statStr := ft.JSONFlowConversationEthernetPath()
```

Comment:
I would prefer to unmarshal the JSON so that the isJSON will be done first and then we can test the object received like this len(obj.nodes) == 0 ...
Because there is not guarantee that the order will be kept